Make endpoint and request options configurable.

robertatcd opened the pull request:
These changes allow the client to be used with other Unleash servers.

My use case is GitLab: I need to be able to specify the feature flags endpoint for a particular project, and I need to pass two headers to identify the environment and the instance.

So in my own `.env` I have:

```
UNLEASH_URL="https://{our GitLab server}"
UNLEASH_FEATURES_ENDPOINT="/api/v4/feature_flags/unleash/{my project id}/features"
```

And in my `config/unleash.php`:

```
  'request_options' => [
    'headers' => [
        'UNLEASH-APPNAME' => 'local',
        'UNLEASH-INSTANCEID' => '{my instance id}',
    ]
  ],
```

robertatcd committed:
Make endpoint and request options configurable.

mikefrancis commented:
Hi @robertatcd, thanks for this.

I'm just wondering how common this feature is?

If this is something very specific to you or your project, it may be better for you to maintain a fork of this.

robertatcd commented:
Hi @mikefrancis 

I found I needed to extend this in order to use the feature flag facility provided by GitLab, which I image will be fairly common.

It uses the same headers as specified in the Unleash docs:

https://docs.getunleash.io/docs/api/client/features

While developing this, I had assumed this was something that was mainly intended for use with servers that host flag sets for multiple projects, and so it would also be required if you were accessing the hosted service run by Unleash themselves:

https://www.unleash-hosted.com/open-source/

But looking at the docs again, I see they're not specific to that hosted service. They almost seem mandatory. So I'm wondering what contexts _don't_ require them? I guess I'm missing something? (I'm new to Unleash and only starting to dabble in feature flags.)

robertatcd commented:
Ah yes, here's the reason I thought the APPNAME and INSTANCEID headers were associated with the hosted Unleash service: they are configurable options in (at least some of) the SDKs mentioned on their site: https://www.unleash-hosted.com/demo/connect/#java

mikefrancis commented:
Ah ok, I see now - thanks for the docs!

I think rather than making the whole request options configurable, what if we set these 2 headers if the respected config options exist?

ppx17 commented:
Explicit options are a bit more in line with how other clients (Java & Go) do it.In addition, the Java client also allows you to add custom headers, in case you host instance behind a proxy that requires auth etc.

In my opinion, the way that is implemented in this PR is still very clear and easy to configure, so I would opt not to add the extra complexity of explicit options and then still needing this to provide potential additional headers.

PS. I'm also a Gitlab user that needs these precise options to use this library.

littleajax commented:
This would be a great addition, coming from the gitlab side as well.

dshafik commented:
I was also looking into this for our multi-application -> unleash setup: +1 here

mikefrancis commented:
My main reservation with the implementation here is that the HTTP is completely separate from the Unleash service singleton and we're starting to see a lot of HTTP things bleeding into the Unleash service.

I think we may need to rethink where the HTTP options are coming from/can be configured.

Other than that, the reasoning makes complete sense - I did wonder why a lot of GitLab people were turning up 😅

robertatcd commented:
(Apologies for not getting back to this until now).

@mikefrancis I'm still leaning towards the headers being configurable options. As @ppx17 commented, it makes them completely configurable. It should cover cases like proxies, other server implementations that need different headers, etc.

For the same reason, I made the API endpoint configurable rather than it being hardcoded to `api/client/features`, as that's needed in GitLab (for one example) to select between multiple projects.

I take your point about all the HTTP stuff being in the Unleash service though. Did you have anything in mind for refactoring it out?

evanshriner commented:
definitely need this endpoint configurable in our setup. +1

robertatcd committed:
Merge branch 'master' into configurable_endpoint_and_options

mikefrancis commented on config/unleash.php:
```diff
@@ -5,6 +5,13 @@
   'url' => env('UNLEASH_URL'),
 
+  // Endpoint for accessing feature flags, on the Unleash server.
+  'features_endpoint' => env('UNLEASH_FEATURES_ENDPOINT', '/api/client/features'),
```
We're using camelCase for the keys in this file. I don't think it's the actual Laravel convention but we should be consistent in this package

robertatcd replied:
Addressed in `ffee84d85bc676ac5309c7a6ca14b5c23f0642b7`

robertatcd committed:
Merge branch 'master' into configurable_endpoint_and_options

robertatcd committed:
fix: camelCase for config settings.

robertatcd committed:
Merge branch 'configurable_endpoint_and_options' of https://github.com/robertatcd/laravel-unleash into configurable_endpoint_and_options

robertatcd committed:
Provide for the use of multiple default request settings

+ move base_uri into requestDefaults config setting
+ use requestDefaults config setting when instantiating the client
+ retain the features endpoint as a separate config setting, in case the client is used to make different requests to different endpoints in the future.

robertatcd committed:
Reduce comment line length.

robertatcd committed:
Update tests to reflect removal of requestOptions config setting.

robertatcd commented:
Wow, where did that month go? Apologies for the long delay again.

@mikefrancis I hope this change is in line with what you were thinking, re: making use of the client initialisation.

I have bundled the `url`/`base_uri` setting into a general `requestDefaults` setting, which is used in the client initialisation.

At the suggestion of a colleague, I refrained from my initial approach of also tacking the features endpoint onto the end of `base_uri`, because that would make the features endpoint specifically the default for the client. His reasoning was that you may want to use that client to access other endpoints on the Unleash server in the future, so it's better to keep that config setting separate, and make use of it in the `->get()` call in `fetchFeatures()`.

mikefrancis commented on src/Client.php:
```diff
@@ -10,9 +10,7 @@ class Client extends GuzzleClient
     public function __construct(Config $config)
     {
         parent::__construct(
-            [
```
I think we may need to do a bit of merging of the config here as by removing the `unleash.url` we are not making this backwards compatible.

I am not sure this feature warrants an entire new major version, so I'm just wondering if there's anything clever we could do to fall back to `unleash.url` if `unleash.requestDefaults` doesn't exist? Does this make sense?

robertatcd replied:
Oh yes, good point.

I have been hit _twice_ this week by unheralded breaking changes, so I'm very sympathetic to that!

Updated to return `unleash.url` to its original spot.

mikefrancis commented:
@robertatcd Tell me about it! Thanks for your latest commits. The last comment is more of a discussion as it would be a breaking change but happy to merge after we've ironed that out.

robertatcd committed:
Return 'url' to its original place in config structure.

robertatcd commented on src/Unleash.php:
```diff
@@ -127,7 +127,7 @@ function () {
     protected function fetchFeatures(): array
     {
-        $response = $this->client->get($this->getFeaturesApiUrl(), $this->getRequestOptions());
+        $response = $this->client->get($this->config->get('unleash.featuresEndpoint'));
```
@mikefrancis I think this would be a breaking change too, actually, if someone used this update without knowing to add `unleash.featuresEndpoint` to their existing config file.

Might be best if I restore the `getFeaturesApiUrl()` function?

robertatcd replied:
Ah, no. Not required, because of course you merge the solution's custom config with the package's config in your `register` method, which means that the default will be used even if the setting is left out of the custom config.

And I've tested and confirmed that theory.

mikefrancis reviewed (approved):
Thank you @robertatcd ❤️

robertatcd commented:
Thanks for the guidance @mikefrancis .

We're now using v0.7 to access GitLab feature flags in production, and it's working well!
